File: Dreams/YearListControl.vb
Imports System.Xml
Imports System.IO
Imports System.Windows.Forms.DataVisualization.Charting

Public Class YearListControl

  Public Event MonthSelected(ByVal Month As String)

  Public Sub LoadYear(ByVal Folder As String, ByVal Year As String)
    Dim intPos As Double = 0.5
    graph.ChartAreas(0).AxisX.CustomLabels.Clear()
    graph.Series.Clear()
    lstDreams.Items.Clear()

    Dim dtYear As DateTime = New DateTime(Year, 1, 1)
    Dim currentMonth As Integer = 0


    Dim objSeries As Series = graph.Series.Add("Year")
    objSeries.ChartType = SeriesChartType.Column
    lblTitle.Text = "Dreams - " & Year

    'objSeries.BorderDashStyle = ChartDashStyle.Solid
    objSeries.Color = Color.FromArgb(200, 65, 140, 240)
    objSeries.BorderWidth = 1
    objSeries("BarLabelStyle") = "Center"


    For Each strMonthFolder As String In Directory.GetDirectories(Folder)

      objSeries.Points.AddY(0)
      objSeries.Points(objSeries.Points.Count - 1).ToolTip = New DirectoryInfo(strMonthFolder).Name

      graph.ChartAreas(0).AxisX.CustomLabels.Add(intPos, intPos + 1, New DirectoryInfo(strMonthFolder).Name)
      intPos += 1

      For Each strDreamFile As String In Directory.GetFiles(strMonthFolder, "*.ld3")
        objSeries.Points(objSeries.Points.Count - 1).YValues(0) += 1
      Next
      Application.DoEvents()
    Next

    Return

    Do

      Dim dtMonth As DateTime = New DateTime(Year, dtYear.Month, 1)

      Dim intPoint As Integer
      Dim lstItem As ListViewItem

      lblTitle.Text = "Dreams - " & Year

      Dim intValue As Integer = 0
      If IO.Directory.Exists(Folder & "\" & Format(dtYear, "MM")) Then


        'objSeries.Label = Format(dtMonth, "MMM")
        intPos = 0.5

        ' Load each dream
        For Each strDreamFile As String In Directory.GetFiles(Folder & "\" & Format(dtYear, "MM"), "*.ld3")
          Dim xmlDream As New XmlDocument
          xmlDream.Load(strDreamFile)
          lstItem = New ListViewItem(xmlDream.DocumentElement.SelectSingleNode("Date").InnerText)
          lstItem.SubItems.Add(xmlDream.DocumentElement.SelectSingleNode("Title").InnerText)
          lstItem.SubItems.Add(xmlDream.DocumentElement.SelectSingleNode("Dream").InnerText.Split(" ").Length)
          lstItem.Tag = strDreamFile
          lstDreams.Items.Add(lstItem)
        Next

        Do

          Dim boolFound As Boolean = False
          For Each lstItem In lstDreams.Items
            If lstItem.Text = dtMonth Then
              boolFound = True
              Exit For
            End If
          Next

          If boolFound Then
            ' intPoint = objSeries.Points.AddY(lstItem.SubItems(2).Text)
            intValue += lstItem.SubItems(2).Text
          Else
            'intPoint = objSeries.Points.AddY(0)
          End If


          dtMonth = dtMonth.AddDays(1)
          intPos = intPos + 1
        Loop Until dtMonth.Month <> dtYear.Month


      End If

      intPoint = objSeries.Points.AddY(intValue)
      objSeries.Points(intPoint).ToolTip = Format(dtYear, "MM")

      dtYear = dtYear.AddMonths(1)
    Loop Until dtYear.Year <> Year
  End Sub

  Private Sub tbcMain_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs)

  End Sub

  Private Sub lstCategories_MouseDown(ByVal sender As System.Object, ByVal e As System.Windows.Forms.MouseEventArgs)

  End Sub

  Private Sub graph_MouseMove(ByVal sender As System.Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles graph.MouseMove
    Try
      Dim result As HitTestResult = graph.HitTest(e.X, e.Y)

      ' Reset Data Point Attributes
      Dim point As DataPoint
      For Each point In graph.Series("Year").Points
        point.Color = Color.FromArgb(200, 65, 140, 240)
      Next point

      If result.ChartElementType = ChartElementType.DataPoint Or result.ChartElementType = ChartElementType.DataPointLabel Or result.ChartElementType = ChartElementType.LegendItem Then
        If result.Series.Name = "Year" Then

          ' Set cursor type 
          Me.Cursor = Cursors.Hand

          ' Find selected data point
          point = graph.Series("Year").Points(result.PointIndex)
          point.Color = Color.FromArgb(255, 65, 140, 240) ' Color.LightSteelBlue
        Else
          ' Set default cursor
          Me.Cursor = Cursors.Default
        End If
      Else
        ' Set default cursor
        Me.Cursor = Cursors.Default
      End If


    Catch ex As Exception

    End Try
  End Sub

  Private Sub graph_MouseDown(ByVal sender As System.Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles graph.MouseDown
    If e.Button = Windows.Forms.MouseButtons.Left Then
      Dim result As HitTestResult = graph.HitTest(e.X, e.Y)
      Dim point As DataPoint
      If result.ChartElementType = ChartElementType.DataPoint Or result.ChartElementType = ChartElementType.DataPointLabel Or result.ChartElementType = ChartElementType.LegendItem Then
        If result.Series.Name = "Year" Then
          ' Set cursor type 
          Me.Cursor = Cursors.Hand

          ' Find selected data point
          point = graph.Series("Year").Points(result.PointIndex)
          RaiseEvent MonthSelected(point.ToolTip)
        End If
      End If
    End If
  End Sub
End Class

File: Dreams/YearListControl.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Xml;
using System.IO;
using System.Windows.Forms.DataVisualization.Charting;
namespace lucidcode.LightenedDream.Dreams
{

	public partial class YearListControl
	{

		public event MonthSelectedEventHandler MonthSelected;
		public delegate void MonthSelectedEventHandler(string Month);

		public void LoadYear(string Folder, string Year)
		{
			double intPos = 0.5;
			graph.ChartAreas[0].AxisX.CustomLabels.Clear();
			graph.Series.Clear();
			lstDreams.Items.Clear();

			DateTime dtYear = new DateTime(Convert.ToInt32(Year), 1, 1);
			int currentMonth = 0;


			Series objSeries = graph.Series.Add("Year");
			objSeries.ChartType = SeriesChartType.Column;
			lblTitle.Text = "Dreams - " + Year;

			//objSeries.BorderDashStyle = ChartDashStyle.Solid
			objSeries.Color = Color.FromArgb(200, 65, 140, 240);
			objSeries.BorderWidth = 1;
			objSeries["BarLabelStyle"] = "Center";



			foreach (string strMonthFolder in Directory.GetDirectories(Folder)) {
				objSeries.Points.AddY(0);
				objSeries.Points[objSeries.Points.Count - 1].ToolTip = new DirectoryInfo(strMonthFolder).Name;

				graph.ChartAreas[0].AxisX.CustomLabels.Add(intPos, intPos + 1, new DirectoryInfo(strMonthFolder).Name);
				intPos += 1;

				foreach (string strDreamFile in Directory.GetFiles(strMonthFolder, "*.ld3")) {
					objSeries.Points[objSeries.Points.Count - 1].YValues[0] += 1;
				}
				Application.DoEvents();
			}

			return;


			do {
				DateTime dtMonth = new DateTime(Convert.ToInt32(Year), dtYear.Month, 1);

				int intPoint = 0;
				ListViewItem lstItem = null;

				lblTitle.Text = "Dreams - " + Year;

				int intValue = 0;

				if (System.IO.Directory.Exists(Folder + "\\" + Strings.Format(dtYear, "MM"))) {

					//objSeries.Label = Format(dtMonth, "MMM")
					intPos = 0.5;

					// Load each dream
					foreach (string strDreamFile in Directory.GetFiles(Folder + "\\" + Strings.Format(dtYear, "MM"), "*.ld3")) {
						XmlDocument xmlDream = new XmlDocument();
						xmlDream.Load(strDreamFile);
						lstItem = new ListViewItem(xmlDream.DocumentElement.SelectSingleNode("Date").InnerText);
						lstItem.SubItems.Add(xmlDream.DocumentElement.SelectSingleNode("Title").InnerText);
						lstItem.SubItems.Add(xmlDream.DocumentElement.SelectSingleNode("Dream").InnerText.Split(' ').Length.ToString());
						lstItem.Tag = strDreamFile;
						lstDreams.Items.Add(lstItem);
					}


					do {
						bool boolFound = false;
						foreach (ListViewItem lstItem_loopVariable in lstDreams.Items) {
							lstItem = lstItem_loopVariable;
              if (lstItem.Text == dtMonth.ToString())
              {
								boolFound = true;
								break; // TODO: might not be correct. Was : Exit For
							}
						}

						if (boolFound) {
							// intPoint = objSeries.Points.AddY(lstItem.SubItems(2).Text)
							intValue += Convert.ToInt32(lstItem.SubItems[2].Text);
						} else {
							//intPoint = objSeries.Points.AddY(0)
						}

						dtMonth = dtMonth.AddDays(1);
						intPos = intPos + 1;
					} while (!(dtMonth.Month != dtYear.Month));


				}

				intPoint = objSeries.Points.AddY(intValue);
				objSeries.Points[intPoint].ToolTip = Strings.Format(dtYear, "MM");

				dtYear = dtYear.AddMonths(1);
			} while (!(dtYear.Year.ToString() != Year));
		}


		private void tbcMain_SelectedIndexChanged(System.Object sender, System.EventArgs e)
		{
		}


		private void lstCategories_MouseDown(System.Object sender, System.Windows.Forms.MouseEventArgs e)
		{
		}

		private void graph_MouseMove(System.Object sender, System.Windows.Forms.MouseEventArgs e)
		{
			try {
				HitTestResult result = graph.HitTest(e.X, e.Y);

				// Reset Data Point Attributes
				DataPoint point = null;
				foreach (DataPoint point_loopVariable in graph.Series["Year"].Points) {
					point = point_loopVariable;
					point.Color = Color.FromArgb(200, 65, 140, 240);
				}

				if (result.ChartElementType == ChartElementType.DataPoint | result.ChartElementType == ChartElementType.DataPointLabel | result.ChartElementType == ChartElementType.LegendItem) {

					if (result.Series.Name == "Year") {
						// Set cursor type 
						this.Cursor = Cursors.Hand;

						// Find selected data point
						point = graph.Series["Year"].Points[result.PointIndex];
						point.Color = Color.FromArgb(255, 65, 140, 240);
						// Color.LightSteelBlue
					} else {
						// Set default cursor
						this.Cursor = Cursors.Default;
					}
				} else {
					// Set default cursor
					this.Cursor = Cursors.Default;
				}



			} catch (Exception ex) {
			}
		}

		private void graph_MouseDown(System.Object sender, System.Windows.Forms.MouseEventArgs e)
		{
			if (e.Button == System.Windows.Forms.MouseButtons.Left) {
				HitTestResult result = graph.HitTest(e.X, e.Y);
				DataPoint point = null;
				if (result.ChartElementType == ChartElementType.DataPoint | result.ChartElementType == ChartElementType.DataPointLabel | result.ChartElementType == ChartElementType.LegendItem) {
					if (result.Series.Name == "Year") {
						// Set cursor type 
						this.Cursor = Cursors.Hand;

						// Find selected data point
						point = graph.Series["Year"].Points[result.PointIndex];
						if (MonthSelected != null) {
							MonthSelected(point.ToolTip);
						}
					}
				}
			}
		}
	}
}
